homeassistant/components/mobile_app/__init__.py

```diff
@@ -131,25 +134,55 @@ async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
     registration_name = f"Mobile App: {registration[ATTR_DEVICE_NAME]}"
     webhook_register(hass, DOMAIN, registration_name, webhook_id, handle_webhook)
 
+    def clean_cloudhook() -> None:
+        """Clean up cloudhook from config entry."""
+        if CONF_CLOUDHOOK_URL in entry.data:
+            data = dict(entry.data)
+            data.pop(CONF_CLOUDHOOK_URL)
+            hass.config_entries.async_update_entry(entry, data=data)
+
+    def on_cloudhook_change(cloudhook: dict[str, Any] | None) -> None:
+        """Handle cloudhook changes."""
+        if cloudhook:
+            if (
+                CONF_CLOUDHOOK_URL in entry.data
+                and entry.data[CONF_CLOUDHOOK_URL] == cloudhook[CONF_CLOUDHOOK_URL]
+            ):
+                return
+
+            hass.config_entries.async_update_entry(
+                entry,
+                data={**entry.data, CONF_CLOUDHOOK_URL: cloudhook[CONF_CLOUDHOOK_URL]},
+            )
+        else:
+            clean_cloudhook()
+
     async def manage_cloudhook(state: cloud.CloudConnectionState) -> None:
         if (
             state is cloud.CloudConnectionState.CLOUD_CONNECTED
             and CONF_CLOUDHOOK_URL not in entry.data
         ):
-            await async_create_cloud_hook(hass, webhook_id, entry)
+            await async_create_cloud_hook(hass, webhook_id)
+        elif (
+            state is cloud.CloudConnectionState.CLOUD_DISCONNECTED
+            and not cloud.async_is_logged_in(hass)
+        ):
+            clean_cloudhook()
+
+    entry.async_on_unload(
+        cloud.async_listen_cloudhook_change(hass, webhook_id, on_cloudhook_change)
+    )
 
     if cloud.async_is_logged_in(hass):
         if (
             CONF_CLOUDHOOK_URL not in entry.data
             and cloud.async_active_subscription(hass)
             and cloud.async_is_connected(hass)
         ):
-            await async_create_cloud_hook(hass, webhook_id, entry)
+            await async_create_cloud_hook(hass, webhook_id)
```

Comment:
As you deleted the storing of the cloudhook in the function, we will not store the hook in `CONF_CLOUDHOOK_URL`. Is this correct?

Reply:
It's stored through the listener on the cloud hook changes. I've made it this way to properly handle the disable/enable of a cloud hook. Otherwise you would need to reload the mobile app integration to enable back a cloud hook.